In [17]:
# Imports and pip installations (if needed)
import pandas as pd
import sklearn

# Part 1: Load the dataset

In [18]:
# Load the given datasets
data_cat = pd.read_csv('chronic_kidney_disease_categorical.csv')
data_num = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Print the data
pd.set_option('display.max_rows', None)  
pd.set_option('display.max_columns', None)  
print(data_cat)
print(data_num)


     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
5       114717  0.0  0.0    normal    normal  notpresent  notpresent   no   
6       950823  4.0  0.0       NaN    normal  notpresent  notpresent   no   
7       901528  0.0  0.0    normal    normal  notpresent  notpresent   no   
8       433613  0.0  0.0    normal    normal  notpresent  notpresent   no   
9       621332  1.0  0.0  abnormal    normal  notpresent  notpresent   no   
10      708255  2.0  0.0    normal    normal  notpresent  notpresent   no   
11      365669  1.0  0.0    normal  abnormal  notpresent  notpresent   no   

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
The dataset contains 25 total attributes, where the label for the data set is "class" which identifies whether the patient has Chronic Kidney Disease (ckd) or notckd. The features in the dataset are both categorical and numerical, where 14 are categorical and 11 are numerical.


Categorical Columns: 
1. rbc (red blood cells): normal, abnormal
2. pc (pus cell): normal, abnormal
3. htn (hypertension): yes, no 


Numerical Columns: 
1. Age: years
2. bp (blood pressure): mm/Hg
3. bgr (Blood Glucose Random): mgs/dl

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [19]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicate_rows_num = data_num[data_num.duplicated()]
total_duplicates_num = duplicate_rows_num.shape[0]
print(f'Total number of duplicate rows in numerical dataset: {total_duplicates_num}')

# Drop duplicate rows
num_dataset_clean = data_num.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicate_rows_cat = data_cat[data_cat.duplicated()]
total_duplicates_cat = duplicate_rows_cat.shape[0]
print(f'total number of duplicate rows in categorical dataset: {total_duplicates_cat}')

cat_dataset_clean = data_cat.drop_duplicates()

Total number of duplicate rows in numerical dataset: 30
total number of duplicate rows in categorical dataset: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:
I believe that all unique IDs are present in both datasets, as we dropped 30 duplicates from both the numerical and categorical datasets, which would bring the total number of rows to 400 for each CSV file. The terminal output of the print function specifies that there are 400 total rows in the merged dataframe, meaning that there were no dropped rows in the process of merging the datasets. 

In [20]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
num_cat_dataset = pd.merge(num_dataset_clean, cat_dataset_clean, on='unique_id', how='outer')

#Print the combined dataset
print(num_cat_dataset)

     unique_id   age     bp    bgr     bu     sc    sod   pot  hemo   pcv  \
0       102677  59.0   70.0   76.0  186.0  15.00  135.0   7.6   7.1  22.0   
1       103570  24.0   80.0  125.0    NaN    NaN  136.0   3.5  15.4  43.0   
2       104675  65.0   80.0  215.0  133.0   2.50    NaN   NaN  13.2  41.0   
3       105923  90.0   90.0  139.0   89.0   3.00  140.0   4.1  12.0  37.0   
4       107297  60.0   70.0    NaN    NaN    NaN    NaN   NaN  16.4  43.0   
5       108785  53.0  100.0  213.0   23.0   1.00  139.0   4.0   NaN   NaN   
6       109053  64.0   60.0  106.0   27.0   0.70  150.0   3.3  14.4  42.0   
7       114717  41.0   80.0  122.0   25.0   0.80  138.0   5.0  17.1  41.0   
8       118191  42.0   70.0   93.0   32.0   0.90  143.0   4.7  16.6  43.0   
9       118596  74.0   60.0   88.0   50.0   0.60  147.0   3.7  17.2  53.0   
10      119638  24.0    NaN  410.0   31.0   1.10    NaN   NaN  12.4  44.0   
11      124304  69.0   70.0  264.0   87.0   2.70  130.0   4.0  12.5  37.0   

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [21]:
# Calculate the percentage of rows that contain atleast one missing value
total_rows = num_cat_dataset.shape[0]

missing_vals = num_cat_dataset[num_cat_dataset.isnull().any(axis=1)]
num_missing_vals = missing_vals.shape[0]

percent_missing = (num_missing_vals / total_rows)*100

# Print %
print(f'Percentage of rows with at least one missing value: {percent_missing:.2f}%')

# Drop these rows from the dataset
num_cat_dataset_clean = num_cat_dataset.dropna()

# Print the Dataset
print(num_cat_dataset_clean)

Percentage of rows with at least one missing value: 61.75%


     unique_id   age     bp    bgr     bu    sc    sod   pot  hemo   pcv  \
0       102677  59.0   70.0   76.0  186.0  15.0  135.0   7.6   7.1  22.0   
6       109053  64.0   60.0  106.0   27.0   0.7  150.0   3.3  14.4  42.0   
7       114717  41.0   80.0  122.0   25.0   0.8  138.0   5.0  17.1  41.0   
8       118191  42.0   70.0   93.0   32.0   0.9  143.0   4.7  16.6  43.0   
9       118596  74.0   60.0   88.0   50.0   0.6  147.0   3.7  17.2  53.0   
11      124304  69.0   70.0  264.0   87.0   2.7  130.0   4.0  12.5  37.0   
13      125587  37.0   60.0  109.0   47.0   1.1  141.0   4.9  15.0  48.0   
14      129053  48.0   80.0  133.0  139.0   8.5  132.0   5.5  10.3  36.0   
15      129654  52.0   80.0  125.0   22.0   1.2  139.0   4.6  16.5  43.0   
16      131563  29.0   80.0   83.0   49.0   0.9  139.0   3.3  17.5  40.0   
18      137148  73.0  100.0  295.0   90.0   5.6  140.0   2.9   9.2  30.0   
24      153227  60.0   90.0  490.0   95.0   2.7  131.0   3.8  11.5  35.0   
25      1551

## Part 3.4: Sort the dataset according to the Labels

In [22]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sort_dataset = num_cat_dataset_clean.sort_values(by='Target')
sort_dataset.reset_index(drop=True, inplace=True)

# Print the dataset
print(sort_dataset)

     unique_id   age     bp    bgr     bu    sc    sod   pot  hemo   pcv  \
0       102677  59.0   70.0   76.0  186.0  15.0  135.0   7.6   7.1  22.0   
1       337834  48.0   70.0  117.0   56.0   3.8  111.0   2.5  11.2  32.0   
2       343710  61.0   80.0  173.0  148.0   3.9  135.0   5.2   7.7  24.0   
3       349892  59.0   70.0  424.0   55.0   1.7  138.0   4.5  12.6  37.0   
4       353025  40.0   70.0  253.0  150.0  11.9  132.0   5.6  10.9  31.0   
5       397388  63.0  100.0  280.0   35.0   3.2  143.0   3.5  13.0  40.0   
6       428486  73.0   70.0   70.0   32.0   0.9  125.0   4.0  10.0  29.0   
7       438182  73.0   80.0  253.0  142.0   4.6  138.0   5.8  10.5  33.0   
8       474407  21.0   90.0  107.0   40.0   1.7  125.0   3.5   8.3  23.0   
9       475200  64.0  100.0  163.0   54.0   7.2  140.0   4.6   7.9  26.0   
10      481293  60.0   60.0  288.0   36.0   1.7  130.0   3.0   7.9  25.0   
11      484175  53.0   90.0   70.0  107.0   7.2  114.0   3.7   9.5  29.0   
12      5147

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [23]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

dummy_dataset = pd.get_dummies(sort_dataset, columns=categorical_columns, drop_first=True)

# Print the dataset
print(dummy_dataset)


     unique_id   age     bp    bgr     bu    sc    sod   pot  hemo   pcv  \
0       102677  59.0   70.0   76.0  186.0  15.0  135.0   7.6   7.1  22.0   
1       337834  48.0   70.0  117.0   56.0   3.8  111.0   2.5  11.2  32.0   
2       343710  61.0   80.0  173.0  148.0   3.9  135.0   5.2   7.7  24.0   
3       349892  59.0   70.0  424.0   55.0   1.7  138.0   4.5  12.6  37.0   
4       353025  40.0   70.0  253.0  150.0  11.9  132.0   5.6  10.9  31.0   
5       397388  63.0  100.0  280.0   35.0   3.2  143.0   3.5  13.0  40.0   
6       428486  73.0   70.0   70.0   32.0   0.9  125.0   4.0  10.0  29.0   
7       438182  73.0   80.0  253.0  142.0   4.6  138.0   5.8  10.5  33.0   
8       474407  21.0   90.0  107.0   40.0   1.7  125.0   3.5   8.3  23.0   
9       475200  64.0  100.0  163.0   54.0   7.2  140.0   4.6   7.9  26.0   
10      481293  60.0   60.0  288.0   36.0   1.7  130.0   3.0   7.9  25.0   
11      484175  53.0   90.0   70.0  107.0   7.2  114.0   3.7   9.5  29.0   
12      5147

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [24]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
outlier_mask = pd.Series([True] * len(dummy_dataset))

for column in numerical_columns:
    mean = dummy_dataset[column].mean()
    std_dev = dummy_dataset[column].std()

    # Define the bounds
    L_bound = mean - 3 * std_dev
    U_bound = mean + 3 * std_dev

    outlier_mask &= (dummy_dataset[column] >= L_bound) & (dummy_dataset[column] <= U_bound)

clean_dataset = dummy_dataset[outlier_mask]
clean_dataset.reset_index(drop=True, inplace=True)

# Print the dataset
print(clean_dataset)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0   
1       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0   
2       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0   
3       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
4       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
5       481293  60.0   60.0  288.0   36.0  1.7  130.0  3.0   7.9  25.0   
6       514721  56.0   90.0  129.0  107.0  6.7  131.0  4.8   9.1  29.0   
7       995177  69.0   70.0  214.0   96.0  6.3  120.0  3.9   9.4  28.0   
8       546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0   
9       980291  62.0   70.0  122.0   42.0  1.7  136.0  4.7  12.6  39.0   
10      955830  55.0   80.0  214.0   73.0  3.9  137.0  4.9  10.9  34.0   
11      923613  64.0   60.0  239.0   58.0  4.3  137.0  5.4   9.5  29.0   
12      843362  71.0   60.0  118.0  12

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [25]:
# Normalize the all Numerical Attributes in the dataset.
for column in numerical_columns: 
    clean_dataset[column] = (clean_dataset[column] - clean_dataset[column].min()) / (clean_dataset[column].max() - clean_dataset[column].min())

# Print the dataset
print(clean_dataset)

     unique_id       age    bp       bgr        bu        sc       sod  \
0       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1       397388  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2       438182  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4       475200  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
5       481293  0.729730  0.00  0.935622  0.169935  0.160494  0.333333   
6       514721  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
7       995177  0.851351  0.25  0.618026  0.562092  0.728395  0.000000   
8       546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
9       980291  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
10      955830  0.662162  0.50  0.618026  0.411765  0.432099  0.566667   
11      923613  0.783784  0.00  0.725322  0.313725  0.481481  0.566667   
12      843362  0.878378  0.00  0.2060

/tmp/ipykernel_51770/3393609942.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_dataset[column] = (clean_dataset[column] - clean_dataset[column].min()) / (clean_dataset[column].max() - clean_dataset[column].min())


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Yes, the unique_id column is unnecessary now that we have the two datasets combined as the unique id would not add any valuable information in terms of a patient's health to the predictive model.

In [26]:
#Remove that column
clean_dataset = clean_dataset.drop(columns=['unique_id'])

# Print the dataset
print(clean_dataset)

          age    bp       bgr        bu        sc       sod       pot  \
0    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.657143   
1    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.171429   
2    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  0.828571   
3    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.171429   
4    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.485714   
5    0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.028571   
6    0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.542857   
7    0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.285714   
8    0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.314286   
9    0.756757  0.25  0.223176  0.209150  0.160494  0.533333  0.514286   
10   0.662162  0.50  0.618026  0.411765  0.432099  0.566667  0.571429   
11   0.783784  0.00  0.725322  0.313725  0.481481  0.566667  0.714286   
12   0.878378  0.00  0.206009  0.751634  0.604938  

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [27]:
# Export the dataframe to a new csv file
clean_dataset.to_csv('clean_dataset.csv', index=False)

# Export the dataframe to a new json file
clean_dataset.to_json('clean_dataset.json', orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here
markdown
| Age | BP | BGR | BU | SC | SOD | POT | Hemo | PCV | WBCC | RBCC | Al | Su | RBC Normal | PC Normal | PCC Present | BA Present | HTN Yes | DM Yes | CAD Yes | Appet Poor | PE Yes | Ane Yes | Target Not CKD |
|---------|---------|-------------|-----------|-----------|-----------|---------|-----------|-----------|------------|------------|----|----|------------|-----------|-------------|------------|---------|--------|---------|-------------|--------|---------|----------------|
| 0.7432 | 0.5 | 0.4421 | 0.9020 | 0.4321 | 0.5 | 0.6571 | 0.1721 | 0.2105 | 0.3952 | 0.1538 | 2 | 0 | false | false | false | false | true | true | true | true | true | true | false |
| 0.7703 | 1.0 | 0.9013 | 0.1634 | 0.3457 | 0.7667 | 0.1714 | 0.6066 | 0.6316 | 0.4435 | 0.4103 | 2 | 2 | true | true | false | true | true | false | true | false | false | false | false |
| 0.9054 | 0.5 | 0.7854 | 0.8627 | 0.5185 | 0.6 | 0.8286 | 0.4016 | 0.4474 | 0.2339 | 0.4359 | 2 | 0 | false | false | false | false | true | true | true | false | false | false | false |
| 0.2027 | 0.75 | 0.1588 | 0.1961 | 0.1605 | 0.1667 | 0.1714 | 0.2213 | 0.1842 | 0.6532 | 0.3333 | 4 | 0 | true | false | true | true | false | false | false | false | false | true | false |
| 0.7838 | 1.0 | 0.3991 | 0.2876 | 0.8395 | 0.6667 | 0.4857 | 0.1885 | 0.2632 | 0.2581 | 0.2051 | 4 | 2 | false | false | false | true | true | true | false | false | true | false | false |
| 0.7297 | 0.0 | 0.9356 | 0.1699 | 0.1605 | 0.3333 | 0.0286 | 0.1885 | 0.2368 | 0.8790 | 0.1026 | 3 | 1 | true | false | true | false | true | false | false | true | false | true | false |
| 0.6757 | 0.75 | 0.2532 | 0.6340 | 0.7778 | 0.3667 | 0.5429 | 0.2869 | 0.3421 | 0.1694 | 0.2051 | 2 | 0 | false | false | false | false | true | false | false | false | false | false | false |
| 0.8514 | 0.25 | 0.6180 | 0.5621 | 0.7284 | 0.0 | 0.2857 | 0.3115 | 0.3158 | 0.5806 | 0.1795 | 4 | 3 | true | false | true | true | true | true | true | false | true | true | false |
| 0.5405 | 0.0 | 0.3991

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

you could mitigate data conversion hallucinations in many ways, such as recucing the number of decimals so the model does not have to account for so many numbers.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [28]:
# Code to rename all the columns in the dataset

column_mapping = {
    'age': 'Age (years)',
    'bp': 'Blood Pressure (mm/Hg)',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell Clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random (mg/dL)',
    'bu': 'Blood Urea (mg/dL)',
    'sc': 'Serum Creatinine (mg/dL)',
    'sod': 'Sodium (mEq/L)',
    'pot': 'Potassium (mEq/L)',
    'hemo': 'Hemoglobin (gms)',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count (cells/cmm)',
    'rbcc': 'Red Blood Cell Count (millions/cmm)',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'class': 'CKD or Not CKD'
}

# Use the rename method to rename columns based on the mapping
clean_dataset.rename(columns=column_mapping, inplace=True)

# Optionally, print the updated DataFrame columns to verify the changes
print(clean_dataset.columns)

Index(['Age (years)', 'Blood Pressure (mm/Hg)', 'Blood Glucose Random (mg/dL)',
       'Blood Urea (mg/dL)', 'Serum Creatinine (mg/dL)', 'Sodium (mEq/L)',
       'Potassium (mEq/L)', 'Hemoglobin (gms)', 'Packed Cell Volume',
       'White Blood Cell Count (cells/cmm)',
       'Red Blood Cell Count (millions/cmm)', 'Albumin', 'Sugar', 'rbc_normal',
       'pc_normal', 'pcc_present', 'ba_present', 'htn_yes', 'dm_yes',
       'cad_yes', 'appet_poor', 'pe_yes', 'ane_yes', 'Target_notckd'],
      dtype='object')


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [29]:

# Converted SQL to Pandas code
print(clean_dataset.columns)
grouped_count = clean_dataset.groupby('Age (years)').size().reset_index(name='count')

# Print the resulting DataFrame
print(grouped_count)

Index(['Age (years)', 'Blood Pressure (mm/Hg)', 'Blood Glucose Random (mg/dL)',
       'Blood Urea (mg/dL)', 'Serum Creatinine (mg/dL)', 'Sodium (mEq/L)',
       'Potassium (mEq/L)', 'Hemoglobin (gms)', 'Packed Cell Volume',
       'White Blood Cell Count (cells/cmm)',
       'Red Blood Cell Count (millions/cmm)', 'Albumin', 'Sugar', 'rbc_normal',
       'pc_normal', 'pcc_present', 'ba_present', 'htn_yes', 'dm_yes',
       'cad_yes', 'appet_poor', 'pe_yes', 'ane_yes', 'Target_notckd'],
      dtype='object')
    Age (years)  count
0      0.000000      1
1      0.081081      1
2      0.121622      1
3      0.148649      1
4      0.189189      1
5      0.202703      1
6      0.216216      1
7      0.229730      4
8      0.243243      1
9      0.256757      2
10     0.297297      1
11     0.310811      2
12     0.324324      5
13     0.351351      2
14     0.364865      4
15     0.378378      3
16     0.391892      2
17     0.405405      1
18     0.418919      2
19     0.432432      2
20  



the code above takes all of the values of the selected column, and counts how many times each number appears in the dataset for the specified column.
